rfd/0003-plugins-unknown-events.md

```diff
@@ -0,0 +1,384 @@
+using JSON and send it to the upstream service. We could send the raw protobuf
+to auth server and have the auth server serialize it to JSON but this is not
+ideal because it will add a lot of overhead to the Auth server and will add
+a lot of latency to the event handler.
+
+Currently, the event handler ignores unknown events.
+
+## Proposal
+
+The proposal consists of having Auth server serve the events as JSON when the
+client requests the JSON endpoints. This will allow the event handler to receive
+the events as JSON and send them to the upstream service without having to
+deserialize the Protobuf and serialize it again to JSON.
+
+To achieve this, we will add two new endpoints to the Auth server that will return
+the events as JSON. The event handler will use these endpoints to retrieve the events
+and session events as JSON.
+
+The new endpoints will be:
+
+```protobuf
+service AuthService {
+    ...
+    // StreamSessionEventsJSON streams audit events from a given session recording in JSON format.
+    // This endpoint is used by the event handler to retrieve the session events as JSON.
+    rpc StreamSessionEventsJSON(StreamSessionEventsRequest) returns (stream EventJSON);
+    // GetEventsJSON gets events from the audit log in JSON format.
+    // This endpoint is used by the event handler to retrieve the events as JSON.
+    rpc GetEventsJSON(GetEventsRequest) returns (EventsJSON);
+}
+
+// EventsJSON represents a list of events.AuditEvent in JSON format.
+message EventsJSON {
+    // Items is a list of typed gRPC formatted audit events.
+    repeated EventJSON Items = 1;
+    // the key of the last event if the returned set did not contain all events found i.e limit <
+    // actual amount. this is the key clients can supply in another API request to continue fetching
+    // events from the previous last position
+    string LastKey = 2;
+}
+
+// EventJSON represents a siggle events.AuditEvent in JSON format.
+message EventJSON {
+    // Type is the type of the event.
+    string Type = 1;
+    // ID is the unique ID of the event.
+    // If the underlying event defines an ID, it will be used, otherwise
+    // ID is a SHA256 hash of the event payload.
+    string ID = 2;
+    // Time is the time when the event was generated.
+    google.protobuf.Timestamp Time = 3 [
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = false
+    ];
+    // Index is the index of the event.
+    int64 Index = 4;
+    // Payload is the JSON event payload.
+    bytes Payload = 5;
+}
+```
+
+Auth server serializes the events to JSON and returns them using `EventJSON` struct.
+The serialization is done using the default event JSON representation for each event type
+besides the `print` event type. The `print` event type is serialized using a custom
+JSON representation because the default JSON representation does not include the
+`data` field which is required by the event handler.
```

Comment:
Again, implementation detail, but almost all if not all audit log backends have events start off as JSON then parsed into a `map[string]any`; it might be possible to avoid roundtripping through the oneof events and just keep the event data unstructured, which would even allow auth servers that don't know about an event to forward it through correctly.

Reply:
done